File: practice/3_AR_model/ar1_model.cpp
#include <TMB.hpp>
template<class Type>
Type objective_function<Type>::operator() ()
{
  // Data
  DATA_VECTOR(y);

  // Parameters
  PARAMETER(log_sd);
  PARAMETER(rho);

  int n_data = y.size();

  // Objective funcction
  Type nll = 0.0;

  //nll -= dnorm(y(0), 0.0, exp(log_sd), true);
  for(int i=1; i<n_data; i++){
    nll -= dnorm(y(i), rho*y[i-1], exp(log_sd)^(1/2), true);
  }

  // Reporting
  Type sd = exp(log_sd);

  REPORT(sd);
  ADREPORT(sd);

  return nll;
}

File: practice/3_AR_model/ar1_model.R
# Roy Burstein
rm(list=ls())

library(TMB)


# need to set to same directory as the template file, also pull from git
# Clone the git directory to your H drive and this should work for anyone
dir <- paste0("/homes/",Sys.info()['user'],"/tmb_transition")
system(paste0('cd ',dir,'\ngit pull origin master'))
setwd(paste0(dir,"/practice/3_AR_model"))

######################
# Simulate data

N       <- 1000 # number of observations
rho     <- 0.9 # temporal autocorrelation
s2      <- .1

# AR data
y    <- numeric(length = N)
y[1] <- rnorm(1,0,s2)
for(i in 2:N) y[i] <- rnorm(1,y[i-1]*rho,sqrt(s2))

# fit in R
arima(y,order=c(1,0,0))


# Fit in TMB
compile("ar1_model.cpp")

# build inputs and object
dyn.load(dynlib("ar1_model"))


obj  <- MakeADFun(data       = list("y"=y),
                  parameters = list("log_sd"=0, 'rho'=0),
                  DLL        = "ar1_model")
